ntoskrnl/se/token.c

```diff
@@ -135,34 +135,235 @@ SepDeleteTokenLock(
     ExFreePoolWithTag(Token->TokenLock, TAG_SE_TOKEN_LOCK);
 }
 
-static NTSTATUS
-SepCompareTokens(IN PTOKEN FirstToken,
-                 IN PTOKEN SecondToken,
-                 OUT PBOOLEAN Equal)
+/**
+ * @brief
+ * Compares the elements of SID arrays provided by tokens.
+ * The elements that are being compared for equality are
+ * the SIDs and their attributes.
+ *
+ * @param[in] SidArrayToken1
+ * SID array from the first token.
+ *
+ * @param[in] CountSidArray1
+ * SID count array from the first token.
+ *
+ * @param[in] SidArrayToken2
+ * SID array from the second token.
+ *
+ * @param[in] CountSidArray2
+ * SID count array from the second token.
+ * 
+ * @return
+ * Returns TRUE if the elements match from either arrays,
+ * FALSE otherwise.
+ */
+static
+BOOLEAN
+SepCompareSidAndAttributesFromTokens(
+    _In_ PSID_AND_ATTRIBUTES SidArrayToken1,
+    _In_ ULONG CountSidArray1,
+    _In_ PSID_AND_ATTRIBUTES SidArrayToken2,
+    _In_ ULONG CountSidArray2)
+{
+    ULONG FirstCount, SecondCount;
+    PAGED_CODE();
+
+    /* Bail out if index counters provided are not equal */
+    if (CountSidArray1 != CountSidArray2)
+    {
+        DPRINT("SepCompareSidAndAttributesFromTokens(): Index counters are not the same!\n");
+        return FALSE;
+    }
+
+    /* Loop over the SID arrays and compare them */
+    for (FirstCount = 0; FirstCount < CountSidArray1; FirstCount++)
+    {
+        for (SecondCount = 0; SecondCount < CountSidArray2; SecondCount++)
+        {
+            if (!RtlEqualSid(SidArrayToken1[FirstCount].Sid,
+                             SidArrayToken2[SecondCount].Sid))
+            {
+                continue;
+            }
+
+            if (SidArrayToken1[FirstCount].Attributes !=
+                SidArrayToken2[SecondCount].Attributes)
+            {
+                continue;
+            }
+
+            break;
```

Comment:
Others may not agree with me, but I feel like this writing is much more clear
```
            PSID_AND_ATTRIBUTES Token1 = &SidArrayToken1[FirstCount];
            PSID_AND_ATTRIBUTES Token2 = &SidArrayToken2[SecondCount];

            if (RtlEqualSid(Token1->Sid, Token2->Sid) && Token1->Attributes == Token2->Attributes)
            {
                break;
            }
```
Same with the second function